src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -0,0 +1,747 @@
+              const n = Math.floor(v / textureSize[i]);
+              v = v - n * textureSize[i];
+              return (n & 1) !== 0 ? textureSize[i] - v - 1 : v;
+            }
+            case 'repeat':
+              return v - Math.floor(v / textureSize[i]) * textureSize[i];
+            default:
+              unreachable();
+          }
+        });
+        const v = load(c);
+        ss.push(v);
+        for (const component of rep.componentOrder) {
+          out[component] = (out[component] ?? 0) + v[component]! * sample.weight;
+        }
+      }
+
+      return out;
+    }
+    case 'textureLoad': {
+      return load(toArray(call.coords!));
+    }
+  }
+}
+
+export async function putDataInTextureThenDrawAndCheckResults<T extends Dimensionality>(
+  device: GPUDevice,
+  texture: Texture,
+  sampler: GPUSamplerDescriptor,
+  calls: TextureCall<T>[]
+) {
+  const results = await doTextureCalls(device, texture, sampler, calls);
+  const errs: string[] = [];
+  const rep = kTexelRepresentationInfo[texture.texels.format];
+  for (let callIdx = 0; callIdx < calls.length; callIdx++) {
+    const call = calls[callIdx];
+    const got = results[callIdx];
+    const expect = expected(call, texture, sampler);
+
+    const gULP = rep.bitsToULPFromZero(rep.numberToBits(got));
+    const eULP = rep.bitsToULPFromZero(rep.numberToBits(expect));
+    for (const component of rep.componentOrder) {
+      const g = got[component]!;
+      const e = expect[component]!;
+      const absDiff = Math.abs(g - e);
+      const ulpDiff = Math.abs(gULP[component]! - eULP[component]!);
+      const relDiff = absDiff / Math.max(Math.abs(g), Math.abs(e));
+      if (ulpDiff > 3 && relDiff > 0.03) {
+        const desc = describeTextureCall(call);
+        errs.push(`component was not as expected:
+      call: ${desc}
+ component: ${component}
+       got: ${g}
+  expected: ${e}
+  abs diff: ${absDiff.toFixed(4)}
+  rel diff: ${(relDiff * 100).toFixed(2)}%
+  ulp diff: ${ulpDiff}
+  sample points:
+`);
+        const expectedSamplePoints = [
+          'expected:',
+          ...(await identifySamplePoints(texture.descriptor, (texels: TexelView) => {
+            return Promise.resolve(
+              expected(call, { texels, descriptor: texture.descriptor }, sampler)
+            );
+          })),
+        ];
+        const gotSamplePoints = [
+          'got:',
+          ...(await identifySamplePoints(
+            texture.descriptor,
+            async (texels: TexelView) =>
+              (
+                await doTextureCalls(device, { texels, descriptor: texture.descriptor }, sampler, [
+                  call,
+                ])
+              )[0]
+          )),
+        ];
+        errs.push(layoutTwoColumns(expectedSamplePoints, gotSamplePoints).join('\n'));
+        errs.push('', '');
+      }
+    }
+  }
+
+  return errs.length > 0 ? new Error(errs.join('\n')) : undefined;
+}
+
+export async function identifySamplePoints(
```

Comment:
Doc comments on functions (at least exported ones) would be very helpful.

Reply:
done. though for this one in particular I just didn't export it.